freqtrade/rpc/emc.py

```diff
@@ -0,0 +1,234 @@
+                                    # We use a lock because it will call RPC methods
+                                    self.handle_producer_message(message)
+
+                            except (asyncio.TimeoutError, websockets.exceptions.ConnectionClosed):
+                                # We haven't received data yet. Check the connection and continue.
+                                try:
+                                    # ping
+                                    ping = await channel.ping()
+
+                                    await asyncio.wait_for(ping, timeout=self.ping_timeout)
+                                    logger.debug(f"Connection to {url} still alive...")
+
+                                    continue
+                                except Exception:
+                                    logger.info(
+                                        f"Ping error {url} - retrying in {self.sleep_time}s")
+                                    await asyncio.sleep(self.sleep_time)
+
+                                    break
+                            except Exception as e:
+                                logger.exception(e)
+                                continue
+                except (
+                    socket.gaierror,
+                    ConnectionRefusedError,
+                    websockets.exceptions.InvalidStatusCode
+                ) as e:
+                    logger.error(f"Connection Refused - {e} retrying in {self.sleep_time}s")
+                    await asyncio.sleep(self.sleep_time)
+
+                    continue
+
+        except asyncio.CancelledError:
+            # Exit silently
+            pass
+
+    def compose_consumer_request(self, type_: str, data: Any) -> Dict[str, Any]:
+        """
+        Create a request for sending to a producer
+
+        :param type_: The RPCRequestType
+        :param data: The data to send
+        :returns: Dict[str, Any]
+        """
+        return {'type': type_, 'data': data}
+
+    # How we do things here isn't set in stone. There seems to be some interest
+    # in figuring out a better way, but we shall do this for now.
+    def handle_producer_message(self, message: Dict[str, Any]):
+        """
+        Handles external messages from a Producer
+        """
+        # Should we have a default message type?
+        message_type = message.get('type', RPCMessageType.STATUS)
+        message_data = message.get('data')
+
+        # We shouldn't get empty messages
+        if message_data is None:
+            return
+
+        logger.debug(f"Received message of type {message_type}")
+
+        # Handle Whitelists
+        if message_type == RPCMessageType.WHITELIST:
+            pairlist = message_data
+
+            # Add the pairlist data to the ExternalPairlist plugin
+            external_pairlist = self._rpc._freqtrade.pairlists._pairlist_handlers[0]
+            external_pairlist.add_pairlist_data(pairlist)
```

Comment:
This is dangerous and bound for failure (and won't support all modes you currently have in mind either).
nothing forces the user to use the externalPairlist provider when in follower mode - and i'm not sure we should.

The way the code currently looks like i could very well use VolumePairlist or staticpairlist - using pairs from the leader where available - and doing the analysis on "this node" where they're not.

assuming the initial pairlist handler is the ExternalPairlist breaks that usecase implicitly, as you'll not be able to have additional data other than pairs with open trades.

Reply:
Should the PairlistManager get the data from the DataProvider, and then append them to the already calculated whitelist?

Reply:
For now we will just add the pairlist data to the dataprovider and allow the strategy to decide what it does with that information.